lib/lineage.js

```diff
@@ -68,75 +73,133 @@ const fetchHydratedDoc = id => {
-
-const collectContactIds = docs => {
+// for data_records, include the first-level contact.
+const collectParentIds = docs => {
   const ids = [];
   docs.forEach(doc => {
-    const contactId = doc.contact && doc.contact._id;
-    if (contactId && !ids.includes(contactId)) {
+    let parent = doc.parent;
+    if (doc.type === 'data_record') {
+      const contactId = doc.contact && doc.contact._id;
+      if (!contactId) {
+        return;
+      }
       ids.push(contactId);
+      parent = doc.contact;
     }
-    let parent = doc.parent;
     while (parent) {
-      if (parent._id && !ids.includes(parent._id)) {
+      if (parent._id) {
         ids.push(parent._id);
       }
       parent = parent.parent;
     }
   });
-  return ids;
+  return _.uniq(ids);
 };
 
-const hydrateDocs = docs => {
-  if (!docs.length) {
-    return Promise.resolve([]);
-  }
-  const ids = collectContactIds(docs);
-  if (!ids.length) {
-    return Promise.resolve(docs);
-  }
+// for data_records, doesn't include the first-level contact (it counts as a parent).
+const collectLeafContactIds = partiallyHydratedDocs => {
+  const ids = [];
+  partiallyHydratedDocs.forEach(doc => {
+    let current = doc;
+    if (current.type === 'data_record') {
+      current = current.contact;
+    }
+    while (current) {
+      const contactId = current.contact && current.contact._id;
+      if (contactId) {
+        ids.push(contactId);
+      }
+      current = current.parent;
+    }
+  });
+  return _.uniq(ids);
+};
+
+const fetchDocs = ids => {
   return new Promise((resolve, reject) => {
+    if (!ids || !ids.length) {
+      return resolve([]);
+    }
     db.medic.fetch({ keys: ids }, (err, results) => {
       if (err) {
         return reject(err);
       }
-      const rows = results.rows;
+      return resolve(results.rows.map(row => row.doc).filter(doc => !!doc));
+    });
+  });
+};
+
+const hydrateParents = (docs, parents) => {
+  if (!parents || !parents.length) {
+    return docs;
+  }
+  docs.forEach(doc => {
+    let current = doc;
+    if (doc.type === 'data_record') {
+      const contactDoc = parents.find(parent => parent._id === current.contact._id);
+      if (contactDoc) {
+        doc.contact = contactDoc;
+      }
+      current = doc.contact;
+    }
 
-      const hydratedDocs = JSON.parse(JSON.stringify(docs));
-      hydratedDocs.forEach(doc => {
-        const row = findRow(rows, doc.contact && doc.contact._id);
-        if (row && row.doc) {
-          doc.contact = row.doc;
-        }
-        let parent = doc;
-        while (parent) {
-          const row = findRow(rows, parent.parent && parent.parent._id);
-          if (row && row.doc) {
-            parent.parent = row.doc;
-          }
-          parent = parent.parent;
+    while (current) {
+      if (current.parent && current.parent._id) {
+        /*jshint loopfunc:true */
+        const parentDoc = parents.find(parent => parent._id === current.parent._id);
```

Comment:
To placate jshint, pull out a function, eg:

```js
const findById = (id, docs) => docs.find(doc => doc._id === id);
```

Or if that feels stupid, add a comment on what "loopfunc" is and why it's not an issue here (and I agree it's not).